Netkan/Services/CachingHttpService.cs

```diff
@@ -0,0 +1,141 @@
+
+            _curl = new CurlEasy { UserAgent = CKAN.Net.UserAgentString };
+
+            var caBundle = ResolveCurlCaBundle();
+            if (caBundle != null)
+            {
+                _curl.CaInfo = caBundle;
+            }
+        }
+
+        public string DownloadPackage(Uri url, string identifier)
+        {
+            EnsureNotDisposed();
+
+            return _cache.GetCachedFilename(url) ?? 
+                _cache.Store(url, CKAN.Net.Download(url), string.Format("netkan-{0}", identifier), move: true);
+        }
+
+        public string DownloadText(Uri url)
+        {
+            EnsureNotDisposed();
+
+            Log.DebugFormat("About to download {0}", url);
+
+            var content = string.Empty;
+
+            _curl.Url = url.ToString();
+            _curl.WriteData = null;
+            _curl.WriteFunction = delegate(byte[] buf, int size, int nmemb, object extraData)
+            {
+                content += Encoding.UTF8.GetString(buf);
+                return size * nmemb;
+            };
+
+            var result = _curl.Perform();
+
+            if (result != CurlCode.Ok)
+            {
+                throw new Exception("Curl download failed with error " + result);
+            }
+
+            Log.DebugFormat("Download from {0}:\n\n{1}", url, content);
+
+            return content;
+        }
+
+        /// <summary>
+        /// Resolves the location of the cURL CA bundle file to use.
+        /// </summary>
+        /// <returns>The absolute file path to the bundle file or null if none is found.</returns>
+        private static string ResolveCurlCaBundle()
+        {
+            const string caBundleFileName = "curl-ca-bundle.crt";
+            const string ckanSubDirectoryName = "CKAN";
+
+            var bundle = new[]
+            {
+                // Working Directory
+                Environment.CurrentDirectory,
+
+                // Executable Directory
+                Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location),
+
+                // %LOCALAPPDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), ckanSubDirectoryName),
+
+                // %APPDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), ckanSubDirectoryName),
+
+                // %PROGRAMDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), ckanSubDirectoryName),
+            }
+            .Select(i => Path.Combine(i, caBundleFileName))
+            .FirstOrDefault(File.Exists);
+
+            Log.InfoFormat("Using curl-ca bundle: {0}",bundle ?? "(none)");
+
+            return bundle;
+        }
+    }
+
+    internal sealed partial class CachingHttpService
+    {
+        private bool _isDisposed;
+
+        ~CachingHttpService()
+        {
+            Dispose(disposing: false);
```

Comment:
> Because during finalization typical reference rules no longer apply. Any managed objects we hold a reference to could already be finalized and references we hold to them would be null, otherwise it would be impossible to garbage collect circular references.

Hooray! This is where my understanding was flawed. My background is in garbage collectors that hold reference counts to be sacred. That means one *can* leak circular references, but it's also guaranteed an object will never be destroyed until its reference count reaches zero¹. (Indeed, one of the purposes of destructors is to break circular references where weak references are inappropriate.)

So in `Dispose()` for .NET we're guaranteed that our managed resources exist (and are still our problem), but in `~CachingHttpService()` all bets are off regarding other references, so we shouldn't touch them.

Thank you *so much* for your patience and explanations; I've levelled up in .NET garbage collection as a result. :)

---

¹ An interesting aside is that at least in Perl, this also means that garbage collection is 100% deterministic, as it happens the moment an object's reference count equals zero, meaning in almost all circumstances destructors play the same role as `Dispose`.